platooning_strategic/src/platoon_strategic.cpp

```diff
@@ -0,0 +1,420 @@
+                
+        platoon_strategic_plugin_discovery_pub_ = nh_->advertise<cav_msgs::Plugin>("plugin_discovery", 1);
+        plugin_discovery_msg_.name = "PlatoonStrategicPlugin";
+        plugin_discovery_msg_.versionId = "v1.0";
+        plugin_discovery_msg_.available = true;
+        plugin_discovery_msg_.activated = false;
+        plugin_discovery_msg_.type = cav_msgs::Plugin::TACTICAL;
+        plugin_discovery_msg_.capability = "tactical_plan/plan_trajectory";
+        
+        pose_sub_ = nh_->subscribe("current_pose", 1, &PlatoonStrategicPlugin::pose_cb, this);
+
+        mob_req_sub_ = nh_->subscribe("incoming_mobility_request_message", 1, &PlatoonStrategicPlugin::mob_req_cb, this);
+        mob_resp_sub_ = nh_->subscribe("incoming_mobility_response_message", 1, &PlatoonStrategicPlugin::mob_resp_cb, this);
+        mob_op_sub_ = nh_->subscribe("incoming_mobility_operation_message", 1, &PlatoonStrategicPlugin::mob_op_cb, this);
+
+        pnh_->param<double>("trajectory_time_length", trajectory_time_length_, 6.0);
+        pnh_->param<std::string>("control_plugin_name", control_plugin_name_, "NULL");
+
+        ros::CARMANodeHandle::setSpinCallback([this]() -> bool {
+            platoon_strategic_plugin_discovery_pub_.publish(plugin_discovery_msg_);
+            return true;
+        });
+
+
+    }
+
+    void PlatoonStrategicPlugin::run_states(){
+        
+        switch (psm_->current_platoon_state)
+        {
+        case PlatoonState::STANDBY:
+            run_standby();
+            break;
+        case PlatoonState::LEADERWAITING:
+            waitingStartTime = ros::Time::now().toSec()*1000;
+            run_leader_waiting();
+            break;
+
+        case PlatoonState::LEADER:
+            run_leader();
+            break;
+        
+        case PlatoonState::CANDIDATEFOLLOWER:
+            candidatestateStartTime = ros::Time::now().toSec()*1000;
+            run_candidate_follower();
+            break;
+        
+        case PlatoonState::FOLLOWER:
+            run_follower();
+            break;
+        
+        default:
+            throw std::invalid_argument("Invalid State");
+            break;
+        }
+
+
+    }
+
+    void PlatoonStrategicPlugin::run()
+    {
+    	initialize();
+        psm_->onMobilityRequestMessage(mobility_req_msg_);
+        psm_->onMobilityOperationMessage(mobility_op_msg_);
+        psm_->onMobilityResponseMessage(mobility_resp_msg_);
+        run_states();
+        ros::CARMANodeHandle::setSpinRate(10.0);
+        ros::CARMANodeHandle::spin();
+
+    }
+
+    void PlatoonStrategicPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+    {
+        pose_msg_ = msg;
+    }
+
+    // TODO: fill this function
+    bool PlatoonStrategicPlugin::plan_trajectory_cb(cav_srvs::PlanTrajectoryRequest &req, cav_srvs::PlanTrajectoryResponse &resp){
+        
+
+        return true;
+    }
+
+
+    void PlatoonStrategicPlugin::run_standby() const {
+
```

Comment:
Empty, if this is used in a spin-loop be sure that we're not busy-waiting on this or we'll spin as fast as possible which'll waste CPU cycles.